### PR TITLE
Fix global state corruption in EditorConfig UI

### LINKED ISSUE #42550: Deprecate public property setters on NotificationOption
See https://github.com/dotnet/roslyn/pull/42323#discussion_r392717268 and https://github.com/dotnet/roslyn/pull/42323#discussion_r393036512 and https://github.com/dotnet/roslyn/pull/42323#discussion_r392718306

### LINKED ISSUE #52802: Microsoft.CodeAnalysis.CSharp.EditorFeatures.UnitTests are extremely flaky
Different, but big number of failures reported on each run. Failed test pass on rerun.

For example, first run:
```
[xUnit.net 01:09:46.78]   Finished:    Microsoft.CodeAnalysis.CSharp.EditorFeatures.UnitTests
========== Test run finished: 24373 Tests (23189 Passed, 1011 Failed, 173 Skipped) run in 70.2 min ==========
```

Rerun of failed tests:
```
[xUnit.net 00:07:38.94]   Finished:    Microsoft.CodeAnalysis.CSharp.EditorFeatures.UnitTests
========== Test run finished: 1011 Tests (1011 Passed, 0 Failed, 0 Skipped) run in 7.7 min ==========
``` 

Typical failures look like this:
```
Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.UseExplicitTypeForConst.UseExplicitTypeForConstTests.TestWithNonConstantTupleType
158ms

 
Error: Actual and expected values differ. Expected shown in baseline of diff:
 class C
 {
     void M()
     {
-        const (int, bool) v = (0, true);
+        const (System.Int32, System.Boolean) v = (0, true);
     }
 }

Expected: True
Actual:   False
```

Another type of failures looks like this. Even though a base line diff is reported, I cannot see what is different between the lines at the beginning of the class declaration. Perhaps the diff is in whitespace/line breaks.
```
Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.UsePatternCombinators.CSharpUsePatternCombinatorsDiagnosticAnalyzerTests.TestOnExpression(expression: "ch < ' ' || ch >= 0x100 || 'a' == ch", expected: "ch is < ' ' or >= (char)0x100 or 'a'")
1s 28ms

 
Error: Actual and expected values differ. Expected shown in baseline of diff:
 
 using System;
 using System.Collections.Generic;
 class C
 {
-    static bool field = ch is < ' ' or >= (char)0x100 or 'a';
-    static bool Method() => ch is < ' ' or >= (char)0x100 or 'a';
-    static bool Prop1 => ch is < ' ' or >= (char)0x100 or 'a';
-    static bool Prop2 { get; } = ch is < ' ' or >= (char)0x100 or 'a';
-    static void If() { if (ch is < ' ' or >= (char)0x100 or 'a') ; }
-    static void Argument1() => Test(ch is < ' ' or >= (char)0x100 or 'a');
-    static void Argument2() => Test(() => ch is < ' ' or >= (char)0x100 or 'a');
-    static void Argument3() => Test(_ => ch is < ' ' or >= (char)0x100 or 'a');
+    static bool field = ch is < ' ' or >= (Char)0x100 or 'a';
+    static bool Method() => ch is < ' ' or >= (Char)0x100 or 'a';
+    static bool Prop1 => ch is < ' ' or >= (Char)0x100 or 'a';
+    static bool Prop2 { get; } = ch is < ' ' or >= (Char)0x100 or 'a';
+    static void If() { if (ch is < ' ' or >= (Char)0x100 or 'a') ; }
+    static void Argument1() => Test(ch is < ' ' or >= (Char)0x100 or 'a');
+    static void Argument2() => Test(() => ch is < ' ' or >= (Char)0x100 or 'a');
+    static void Argument3() => Test(_ => ch is < ' ' or >= (Char)0x100 or 'a');
     static void Test(bool b) {}
     static void Test(Func<bool> b) {}
     static void Test(Func<object, bool> b) {}
-    static void For() { for (; ch is < ' ' or >= (char)0x100 or 'a'; ); }
-    static void Local() { var local = ch is < ' ' or >= (char)0x100 or 'a'; }
-    static void Conditional() { _ = ch is < ' ' or >= (char)0x100 or 'a' ? ch is < ' ' or >= (char)0x100 or 'a' : ch is < ' ' or >= (char)0x100 or 'a'; }
-    static void Assignment() { _ = ch is < ' ' or >= (char)0x100 or 'a'; }
-    static void Do() { do ; while (ch is < ' ' or >= (char)0x100 or 'a'); }
-    static void While() { while (ch is < ' ' or >= (char)0x100 or 'a') ; }
-    static bool When() => o switch { _ when ch is < ' ' or >= (char)0x100 or 'a' => ch is < ' ' or >= (char)0x100 or 'a' };
-    static bool Return() { return ch is < ' ' or >= (char)0x100 or 'a'; }
-    static IEnumerable<bool> YieldReturn() { yield return ch is < ' ' or >= (char)0x100 or 'a'; }
-    static Func<object, bool> SimpleLambda() => o => ch is < ' ' or >= (char)0x100 or 'a';
-    static Func<bool> ParenthesizedLambda() => () => ch is < ' ' or >= (char)0x100 or 'a';
-    static void LocalFunc() { bool LocalFunction() => ch is < ' ' or >= (char)0x100 or 'a'; }
+    static void For() { for (; ch is < ' ' or >= (Char)0x100 or 'a'; ); }
+    static void Local() { var local = ch is < ' ' or >= (Char)0x100 or 'a'; }
+    static void Conditional() { _ = ch is < ' ' or >= (Char)0x100 or 'a' ? ch is < ' ' or >= (Char)0x100 or 'a' : ch is < ' ' or >= (Char)0x100 or 'a'; }
+    static void Assignment() { _ = ch is < ' ' or >= (Char)0x100 or 'a'; }
+    static void Do() { do ; while (ch is < ' ' or >= (Char)0x100 or 'a'); }
+    static void While() { while (ch is < ' ' or >= (Char)0x100 or 'a') ; }
+    static bool When() => o switch { _ when ch is < ' ' or >= (Char)0x100 or 'a' => ch is < ' ' or >= (Char)0x100 or 'a' };
+    static bool Return() { return ch is < ' ' or >= (Char)0x100 or 'a'; }
+    static IEnumerable<bool> YieldReturn() { yield return ch is < ' ' or >= (Char)0x100 or 'a'; }
+    static Func<object, bool> SimpleLambda() => o => ch is < ' ' or >= (Char)0x100 or 'a';
+    static Func<bool> ParenthesizedLambda() => () => ch is < ' ' or >= (Char)0x100 or 'a';
+    static void LocalFunc() { bool LocalFunction() => ch is < ' ' or >= (Char)0x100 or 'a'; }
     static int i;
     static int? nullable;
     static object o;
     static char ch;
 }
 

Expected: True
Actual:   False
```

I am not sure what is going on here. One theory is, perhaps, the tests are some how depended, i.e. relying on some shared data (whatever that might be) and there is a "race" to initialize/modify the data. There might be some dependency on environment, etc.

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #52802
Closes #42550

### CODE REVIEW

**Review by jmarolf — APPROVED** [review]

File: src/EditorFeatures/Core/EditorConfigSettings/Data/CodeStyle/CodeStyleSetting.PerLanguageEnumCodeStyleSetting.cs [hunk 1]
```diff
@@ -37,16 +37,14 @@ private class PerLanguageEnumCodeStyleSetting<T> : EnumCodeStyleSettingBase<T>
 
             protected override void ChangeSeverity(NotificationOption2 severity)
             {
-                var option = GetOption();
-                option.Notification = severity;
-                _ = Updater.QueueUpdateAsync(_option, option);
+                ICodeStyleOption option = GetOption();
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/EditorConfigSettings/Data/CodeStyle/CodeStyleSetting.PerLanguageEnumCodeStyleSetting.cs:40]: var?

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/EditorFeatures/Core/EditorConfigSettings/Data/CodeStyle/CodeStyleSetting.PerLanguageEnumCodeStyleSetting.cs (hunk 1 above)
Comment (sharwell) [reply src/EditorFeatures/Core/EditorConfigSettings/Data/CodeStyle/CodeStyleSetting.PerLanguageEnumCodeStyleSetting.cs:40]: Not possible. The next line is using an explicitly implemented interface method.

**Comment by Youssef1313** [comment]
@sharwell the issues didn't get closed as the target branch isn't `main`.

**Comment by Cosifne** [comment]
> @sharwell the issues didn't get closed as the target branch isn't `main`.

@Youssef1313 
This fix needs to go to 16.10 first. We have an auto pipeline to flow changes to main.

**Comment by Youssef1313** [comment]
> > @sharwell the issues didn't get closed as the target branch isn't `main`.
> 
> @Youssef1313
> This fix needs to go to 16.10 first. We have an auto pipeline to flow changes to main.

@Cosifne Yeah I know they will flow to `main`. But I think this won't close the linked issues?

**Comment by Cosifne** [comment]
> > > @sharwell the issues didn't get closed as the target branch isn't `main`.
> > 
> > 
> > @Youssef1313
> > This fix needs to go to 16.10 first. We have an auto pipeline to flow changes to main.
> 
> @Cosifne Yeah I know they will flow to `main`. But I think this won't close the linked issues?

Do you mean #42550?

**Comment by Cosifne** [comment]
My understanding is this PR has sealed the CodeStyleOption & mark it as obsolete in this PR.
Anything else is needed?
Or do you mean this should be fixed by the PR which flows the change to main? @Youssef1313

**Comment by Youssef1313** [comment]
> My understanding is this PR has sealed the CodeStyleOption & mark it as obsolete in this PR.
> Anything else is needed?
> Or do you mean this should be fixed by the PR which flows the change to main? @Youssef1313

I meant that both https://github.com/dotnet/roslyn/issues/42550 and https://github.com/dotnet/roslyn/issues/52802 should be closed (usually this is done automatically, but GitHub doesn't do that when PRs doesn't target the default branch).

But @sharwell has done that, so nothing is needed now.

**Pull request merged**